### DISCUSSION TITLE
FPGA Reset button

### DISCUSSION THREAD

**Discussion by pottendo** [zeldin/OrangeCart#10] [Ideas]
Hi,
is it possible to add some way to reset the FPGA (restart the bitstream) without disconnect of the USB cable or power-off of the C64?

Rationale: if I want to load some SW (e.g. on RVCop64) via litex_term over USB, I need to reboot the bitstream first. Currently this means I need to power off by unplugging the USB cable and turning off the C64.
As I did this _very_ often already, I feel that my USB connector on the OC is already getting a bit unreliable.

I was thinking that a pushbutton e.g. on the OC GPIOs or a 'long-press' on the existing pushbutton could do the trick with some support of the bitstream?

What do you think?
pottendo

**Comment by zeldin** [comment]
Yes, pulling pin M8 (`rst_n` in LiteX) low will reset the FPGA.  It should first go to the bootloader, and if the pushbutton is pressed down stay there, otherwise proceed to the main bitstream.  The reset signal is also available on the JTAG connector.

Personally I prefer unplugging the USB cable at the other end.  :smile:

**Reply by pottendo** [reply]
ad JTAG: does this mean that I just need to connect the SRST pin (J6 pin) to GND? 
(https://github.com/zeldin/OrangeCart/wiki/Connectors, the FPGA pin is R9, not M8?) 
e.g. by soldering a suitable push-button/cable?
Still I'd prefer to route this M8 pin to one of the GPIOs - then there's no danger to kill the bootloader by wobbly cabling or other deficiencies of unskilled guys like me wrt. HW stuff.

I don't have a clue how to work things like this into litex - so I may use your approach, where to disconnect. ;-)  unless you provide some guides... :-D

bye, pottendo

**Reply by zeldin** [reply]
> ad JTAG: does this mean that I just need to connect the SRST pin (J6 pin) to GND?

Yes, that is correct.  There is a GND pin opposite to it, so you could even use the classic paperclip method.  😸  (Not really recommended as it's not ESD safe.)

> (the FPGA pin is R9, not M8?)

M8 and R9 are connected by a PCB trace.  R9 is the actual reset pin, but it can't be controlled from inside the FPGA so I have hooked it up to a normal controllable pin (i.e. M8) to allow for the FPGA to reset itself.

> I don't have a clue how to work things like this into litex

Well, the simplest way to have M8 controlled by, say, GPIO:1, would be to just do
```
        platform.add_extension([("my_btn", 0, Pins("GPIO:1"))])
        self.comb += platform.request("rst_n").eq(platform.request("my_btn"))
```
at the end of the `__init__` function of `BaseSoC` in `basesoc.py`.  (You would also need to add
```
from litex.build.generic_platform import Pins
```
at the top of the file.)  But driving M8 as push/pull has the disadvantage that it prevents SRST in the JTAG connector from working when your bitstream is running.  So a better way is to do like the bootloader does it, and instead of the `self.comb` fragment above do this:
```
        ts = TSTriple()
        self.specials += ts.get_tristate(platform.request("rst_n"))
        self.comb += [
            ts.o.eq(0),
            ts.oe.eq(~platform.request("my_btn"))
        ]
```
This makes M8 tri-stated when the GPIO button is not pushed, allowing reset from the JTAG connector to work as well. (Here
```
from migen import TSTriple
```
would also be needed.)

Note that if you want to connect a pushbutton to GPIO:1, you will need to provide your own pull-up resistor to 3.3V, since the GPIO pins do not have any pull-up or pull-down installed by default.  On the M8/R9/SRST side there is already such a resistor so if you use SRST then just a button shorting to GND would be sufficient.

**Reply by pottendo** [reply]
...looks very promising! Thanks!
How would you dimension this pull-up resistor? The pull-up could be connected to J5, Pin6, right?
It even sounds like a feature you may want to integrate (as an option) into the official bitstreams... ;-)

**Reply by zeldin** [reply]
Pin 6 has 3.3V, yes. Anything from 4.7k (the pull-up used for the current push-button) to 100k (the pull-up used for SRST) should work.

Well, the point of the GPIO pins is that they don't have any predetermined function, so... Using long-push on the existing button to re-enter the bootloader is something I might consider as a standard feature though.

**Reply by pottendo** [reply]
the long-push would be of course the best solution (no blockage of the GPIO and no HW fiddeling needed) - I thought this may be more complex to solve...
